Purchases/Caching/RCDeviceCache.m

```diff
@@ -201,29 +227,80 @@ - (RCSubscriberAttributeDict)storedSubscriberAttributesForAppUserID:(NSString *)
+                                                         ?: @{}.mutableCopy;
+    NSMutableDictionary *mutableSubscriberAttributes = subscriberAttributes.mutableCopy;
+    [mutableSubscriberAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+
+    allSubscriberAttributes[appUserID] = mutableSubscriberAttributes;
+    [self.userDefaults setObject:allSubscriberAttributes forKey:RCSubscriberAttributesKey];
+
+    NSString *legacyAttributesKey = [self legacySubscriberAttributesCacheKeyForAppUserID:appUserID];
+    [self.userDefaults removeObjectForKey:legacyAttributesKey];
+}
+
+- (nullable NSDictionary *)valueForLegacySubscriberAttributes:(NSString *)appUserID {
+    return [self.userDefaults dictionaryForKey:[self legacySubscriberAttributesCacheKeyForAppUserID:appUserID]];
+}
+
+- (NSString *)legacySubscriberAttributesCacheKeyForAppUserID:(NSString *)appUserID {
     NSString *attributeKey = [NSString stringWithFormat:@"%@", appUserID];
-    return [RCSubscriberAttributesKeyBase stringByAppendingString:attributeKey];
+    return [RCLegacySubscriberAttributesKeyBase stringByAppendingString:attributeKey];
 }
 
-#pragma mark - private methods
+- (NSDictionary<NSString *, NSDictionary *> *)storedAttributesForAllUsers {
+    return [self.userDefaults dictionaryForKey:RCSubscriberAttributesKey] ?: @{};
+}
 
-- (void)clearOfferingsCache {
-    [self.offeringsCachedObject clearCache];
+- (NSDictionary<NSString *, RCSubscriberAttributeDict> *)unsyncedAttributesForAllUsers {
+    NSDictionary *attributesDict = self.storedAttributesForAllUsers;
+    NSMutableDictionary *attributes = [[NSMutableDictionary alloc] init];
+
+    for (NSString *appUserID in attributesDict.allKeys) {
+        NSDictionary *attributesDictForUser = (NSDictionary *) attributesDict[appUserID];
+        NSMutableDictionary *attributesForUser = [[NSMutableDictionary alloc] init];
+
+        for (NSString *attributeKey in attributesDictForUser.allKeys) {
+            NSDictionary *attributeDict = (NSDictionary *) attributesDictForUser[attributeKey];
+            RCSubscriberAttribute *attribute = [[RCSubscriberAttribute alloc] initWithDictionary:attributeDict];
+            if (!attribute.isSynced) {
+                attributesForUser[attributeKey] = attribute;
+            }
+        }
+        attributes[appUserID] = attributesForUser;
+    }
+    return attributes;
 }
 
-- (NSString *)purchaserInfoUserDefaultCacheKeyForAppUserID:(NSString *)appUserID {
-    return [RCPurchaserInfoAppUserDefaultsKeyBase stringByAppendingString:appUserID];
+- (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
+    @synchronized (self) {
+        if ([self numberOfUnsyncedAttributesForAppUserID:appUserID] != 0) {
+            return;
+        }
+        NSMutableDictionary <NSString *, NSDictionary *>
+            *groupedAttributes = self.storedAttributesForAllUsers.mutableCopy;
+        [groupedAttributes removeObjectForKey:appUserID];
```

Comment:
We are removing all attributes for that user id. We can be sure they have been synced right?

Reply:
they should be - we check that `numberOfUnsyncedAttributesForAppUserID == 0` a couple of lines before doing this